dali/operators/reader/loader/webdataset/tar_utils.cc

```diff
@@ -0,0 +1,228 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/webdataset/tar_utils.h"
+#include <libtar.h>
+#include <algorithm>
+#include <cstdarg>
+#include <cstdlib>
+#include <list>
+#include <string>
+#include <tuple>
+#include <utility>
+#include "dali/core/error_handling.h"
+#include "dali/core/math_util.h"
+#include "dali/core/util.h"
+
+namespace dali {
+namespace detail {
+
+namespace {
+
+constexpr size_t kBlockSize = T_BLOCKSIZE;
+static_assert(is_pow2(kBlockSize),
+              "The implementation assumes that the block size is a power of 2");
+
+constexpr uint64_t kEmptyEofBlocks = 2;
+constexpr uint64_t kTarArchiveBufferInitSize = 1;
+
+std::mutex instances_mutex;
+std::list<std::vector<TarArchive*>> instances_registry = {
+    std::vector<TarArchive*>(kTarArchiveBufferInitSize)};
+TarArchive** instances = instances_registry.back().data();
+
+int Register(TarArchive* archive) {
+  std::lock_guard<std::mutex> instances_lock(instances_mutex);
+  for (auto& instances_entry : instances_registry.back()) {
+    if (instances_entry != nullptr) {
+      continue;
+    }
+    instances_entry = archive;
+    return &instances_entry - instances;
+  }
+  std::vector<TarArchive*>& old = instances_registry.back();
+  instances_registry.emplace_back();
+  std::vector<TarArchive*>& curr = instances_registry.back();
+  curr.reserve(old.size() * 2);
+  curr = old;
+  curr.push_back(archive);
+  curr.resize(old.size() * 2, nullptr);
+  instances = curr.data();
+  return old.size();
+}
+
+inline void Unregister(int instance_handle_) {
+  instances[instance_handle_] = nullptr;
+}
+
+inline TAR* ToTarHandle(void* handle) {
+  return reinterpret_cast<TAR*>(handle);
+}
+
+inline TAR** ToTarHandle(void** handle) {
+  return reinterpret_cast<TAR**>(handle);
+}
```

Comment:
It is a matter of taste but it is not obvious to me that these wrappers add any value. In places where they are used they obfuscate a fact that this is just a change of pointer type. Maybe you can consider removing them?

Reply:
They are here to avoid lots of reinterpret_cast

Reply:
@awolant It's not perfect, but probably better than having casts all over the place. Ideally we'd forward-declare `TAR`, but it has a C-style declaration which in C++ becomes an anonymous struct which cannot be forward declared:
```
typedef struct {... } TAR;
```